### DISCUSSION TITLE
Chapter 1 - require JS for who use React

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#1] [Chapter1]
# 1. JS Equality comparisons

#### JS Data type
+ boolean
+ null[^1]
+ undefined[^2]
+ number
+ string
+ symbol
+ bigint
+ object

JS has 8 types of data. But we are not going to talk about all type rather check what we need. From a broad perspective, We can classify into two kind of 8 types. One is `primitive type` (1 ~ 7)  the other is `object` (8). One of the most different between both is the how to save values. `primitive type`s save value as a value. `object` type saves a value as a reference. I hate references. Before we skip to talk about reference's detail, Two points we should know.

#### references are not a value as it selves, It's an addresses. So...
- You can't change part of them.
- You can't comparison as you want. even if they are look same each other.

Especially In React, The second one is important for us. `props` is an `object type`. So we can't compare unless do something. And React did it for us. `shallowEqual`, it a function that can make comparison between objects. But It's not perfect, So it only can make comparison with 1 deps between two object. In fact, React made it like this on purpose. Comparison into deep can make low performance.

Here is the key point of Chapter 1.
 - ### ***Do not set an object as a value inside of props. If you do, It doesn't guarantee your components are re-render properly.***

![image](https://github.com/honaseung/react-deep-dive/assets/53066347/e7037d9c-f600-4d66-b55f-cc0d958c95b7)

_* It's like a finding different spot between images._

> [!NOTE]
New Terms
Polyfill: Kind of codes that support old browsers which doesn't have trendy function.

[^1]: presenting of blank of value. like a zero in number.
[^2]: presenting of variable is declared but haven't assigned yet.

**Comment by honaseung** [comment]
As I know, JS treat function as an object as well. Is it ok to pass a function through props?

**Comment by chaminju927** [comment]
I don't get exactly what "Do not set an object as a value inside of props." means. Is there any detail examples about  "object as a value inside of props"  you can provide?

**Reply by honaseung** [reply]
It means React only can compare props depth level 1. So you must not use object as a value of props. In JS, Objects are not same even if both looks same. Here is an example.

```
const obj1 = { a: 1, b: 2 };
const obj2 = { a: 1, b: 2 };
ob1 === obj2 // false
```

Object type is a reference type instead of primitive. So JS saves reference at variable. If you save objects as a value of props, React can not compare props properly.
